### DISCUSSION TITLE
Why does `spawn` for a single thread runtime require `Send` ?

### DISCUSSION THREAD

**Discussion by tsuberim** [tokio-rs/tokio#4998] [Q&A] [answered]
I have a non-`Send` type that I want to pass across `await`, I know that this is impossible when we cross thread boundaries but why should it be impossible when I specifically created a single-threaded runtime using `new_current_thread(..)` ?

**Comment by Darksonn** [comment, marked as answer]
The `Runtime` object itself is `Send`, so you can still send it across threads. Use a `LocalSet` if you want non-Send tasks.

**Reply by tsuberim** [reply]
Hmm... because the runtime keeps a state of the current futures? So if a future contains a ref to a non-send that makes the entire `Runtime` non-sendable. Is this correct?

Maybe we can provide a special `SingleThreadRuntime` without those restrictions.

**Reply by Noah-Kennedy** [reply]
The runtime is `Send` and `Sync` to enable you to do things like access it from multiple threads simultaneously. `LocalSet` provides a way to have `!Send` futures in the root task. This works because the root task is entered into via block_on, so the runtime can't be sent while a localset is in operation.

**Reply by Darksonn** [reply]
> Maybe we can provide a special `SingleThreadRuntime` without those restrictions.

We do, and it's called `LocalSet`.
